core/src/main/java/org/elasticsearch/index/mapper/DocumentMapperParser.java

```diff
@@ -172,13 +176,121 @@ private static String getRemainingFields(Map<?, ?> map) {
             // if we don't have any keys throw an exception
             throw new MapperParsingException("malformed mapping no root object found");
         }
+        String actualType;
+        Map<String, Object> mapping;
+
         String rootName = root.keySet().iterator().next();
-        Tuple<String, Map<String, Object>> mapping;
         if (type == null || type.equals(rootName)) {
-            mapping = new Tuple<>(rootName, (Map<String, Object>) root.get(rootName));
+            actualType = rootName;
+            mapping = (Map<String, Object>) root.get(rootName);
         } else {
-            mapping = new Tuple<>(type, root);
+            actualType = type;
+            mapping = root;
+        }
+
+        mapping = removeDotsInFieldNames(mapping);
+        return new Tuple<>(actualType, mapping);
+    }
+
+    static Map<String,Object> removeDotsInFieldNames(Map<String, ?> mapping) {
+        Map<String, Object> result = new HashMap<>(mapping);
+        Object propertiesObject = result.get(PROPERTIES_KEY);
+        if (propertiesObject instanceof Map) {
+            @SuppressWarnings("unchecked")
+            Map<String, ?> properties = (Map<String, ?>) propertiesObject;
+            properties = removeDotsInFieldNamesFromProperties(properties);
+            result.put(PROPERTIES_KEY, properties);
+        }
```

Comment:
Shouldn't this always be a Map? Should we throw an exception (or assert) if this check fails?